### ISSUE TITLE
InlinedVector::clear() should lazily avoid deallocating out-of-line memory

### ISSUE THREAD

**Issue by udi-barzilai** [abseil/abseil-cpp#776]
abseil-cpp/absl/container/inlined_vector.h 

**Describe the bug**

`InlinedVector::clear()` method should lazily avoid deallocating out-of-line memory, if already allocated.
This would be in line with the behavior of `std::vector`, for which it is designed to be a drop-in replacement.
Deallocation on `clear()` means that if any number of elements beyond the inline capacity are re-pushed into the container, it will reallocate and re-copy contents unnecessarily.
As of this writing, the master branch still has `storage_.DeallocateIfAllocated();` in `clear()`

**Steps to reproduce the bug**

    InlinedVector<int, 8> container;
    container.reserve(max_size); // because I KNOW what the maximum required size is, but not necessarily at compile time
    for (int count=max_size; count>0; --count) {
    #if UGLY_WORKAROUND
        container.erase(container.begin(), container.end());  // does not deallocate
    #else
        container.clear();  // deallocates storage :-(
    #endif
        for (int i=0; i<count; ++i)
            container.push_back(i);
        do_something_with(container);
    }

If called with `max_size` set to 16, the above code will trigger rellocation+copy at least 8 times.
With the workaround (or correct `clear()` behavior) it will never reallocate/copy.

**What version of Abseil are you using?**
Any, AFAIK. 
Specifically, `lts_2020_02_25` has it.

**What operating system and version are you using**
Irrelevant, but Ubuntu 18.04

**What compiler and version are you using?**
g++ 7.5.0

**What build system are you using?**
None (make)

**Comment by derekmauro** [comment]
@CJ-Johnson looked into making this change a year ago, but found it broke a lot of code and was unable to make the change.

@CJ-Johnson, are you willing to take another shot at this?

**Comment by despite567** [comment]
Tap on a clip to paste it in the text box.Touch and hold a clip to pin it. Unpinned clips will be deleted after 1 hour.Use the edit icon to pin, add or delete clips.Welcome to Gboard clipboard, any text you copy will be saved here.

**Comment by CJ-Johnson** [comment]
I don't have the cycles at the moment, but it is on my list!

For tracking purposes, the corresponding bug ID is 138307414

**Comment by jakra-mb** [comment]
This is a serious reason for us to avoid InlinedVector. In a lot of cases there are loops over many elements, and inside the loop we call functions to do something and return the result in a vector. We can keep that vector around outside the loop to reuse it each time. With std::vector, it will just grow to the maximum size needed and then stay there. But this still requires allocations. We'd like to get rid of the allocations as much as possible, so InlinedVector is a good fit. But since clear() always resets the capacity, it may now end up doing much more allocations than std::vector would have done and only the case where the inline capacity is fully sufficient for all cases results in no allocations.

We've done performance measurements, and unfortunately, InlinedVector often performs worse than std::vector, simply because of this.

Looking at the code, it looks to me that if `clear()` was simply implemented as `resize(0)` that should already be it, no?

**Comment by derekmauro** [comment]
> Looking at the code, it looks to me that if clear() was simply implemented as resize(0) that should already be it, no?

The problem is not fixing InlinedVector. The problem is fixing the code that is somehow dependent on the current behavior. No one currently has cycles for that.

**Comment by jtorresfabra** [comment]
Hey @derekmauro Thanks for answering, could you ellaborate a bit on what code depends on this behaviour? Do you mean in the abseil lib or in client code?

**Comment by derekmauro** [comment]
> Hey @derekmauro Thanks for answering, could you ellaborate a bit on what code depends on this behaviour? Do you mean in the abseil lib or in client code?

It is customer code in our internal codebase.

**Comment by EricGriffith** [comment]
If there are no cycles to correct this, can an additional clear method be added that has the standard library behavior?

**Comment by derekmauro** [comment]
> If there are no cycles to correct this, can an additional clear method be added that has the standard library behavior?

No. If you need this today, you can do `iv.erase(iv.begin(), iv.end());`.

**Comment by derekmauro** [comment]
I tracked down the code that was depending on the deallocating behavior and fixed it, and I have now changed `clear()` to not deallocate. If no one comes out to complain in the next couple days I will call this fixed.

**Comment by udi-barzilai** [comment]
@derekmauro Thanks for taking the time to do this.

If you can call it fixed by this Thursday, I would love to have this issue closed exactly on its 5th anniversary 😃

**Comment by derekmauro** [comment]
Going to marked this one as fixed a day early 😄

**Comment by CJ-Johnson** [comment]
Thanks for tackling this, @derekmauro

**Comment by jakra-mb** [comment]
Just saw your fix. Note that I tried the same fix in a custom copy. It works up until you use a type without a default constructor in it (ie classes that can only be constructed with arguments). Then the resize(0) call requires the constructor to exist (even though it's not called) and won't compile.

My solution was to add a `Clear` function to the Storage object, which is just a trimmed down version of the resize function with everything removed that would go through the allocation code paths.
